File: src/Scripting/Perl/observer.pl
#!/usr/bin/env perl
use strict;
use warnings;

my %subscribers;

sub subscribe {
    my ($name, $handler) = @_;
    return 0 if exists $subscribers{$name};
    $subscribers{$name} = $handler;
    return 1;
}

sub unsubscribe {
    my ($name) = @_;
    return 0 unless exists $subscribers{$name};
    delete $subscribers{$name};
    return 1;
}

sub publish {
    my ($event) = @_;
    $_->($event) for values %subscribers;
}

my (@audit_log, @dashboard_log);
my $audit = sub { push @audit_log, shift };
my $dashboard = sub { push @dashboard_log, shift };

sub example_passes {
    %subscribers = ();
    @audit_log = ();
    @dashboard_log = ();

    return 0 unless subscribe('audit', $audit);
    return 0 unless subscribe('dashboard', $dashboard);
    return 0 if subscribe('audit', $audit);

    publish('created');
    return 0 unless join(',', @audit_log) eq 'created';
    return 0 unless join(',', @dashboard_log) eq 'created';

    return 0 unless unsubscribe('dashboard');
    return 0 if unsubscribe('dashboard');

    publish('approved');
    return 0 unless join(',', @audit_log) eq 'created,approved';
    return 0 unless join(',', @dashboard_log) eq 'created';

    return 1;
}

if (example_passes()) {
    print "OBSERVER_PERL_OK\n";
    exit 0;
}

die "Observer contract failed\n";

File: src/DataScience/Julia/observer.jl
module ObserverExample

mutable struct Subject
    observers::Dict{Symbol, Function}
end

Subject() = Subject(Dict{Symbol, Function}())

function subscribe!(subject::Subject, name::Symbol, observer::Function)
    haskey(subject.observers, name) && return false
    subject.observers[name] = observer
    true
end

function unsubscribe!(subject::Subject, name::Symbol)
    haskey(subject.observers, name) || return false
    delete!(subject.observers, name)
    true
end

function publish(subject::Subject, id::Int)
    Dict(name => observer(id) for (name, observer) in subject.observers)
end

function example_passes()
    subject = Subject()
    subscribe!(subject, :audit, id -> "audit:$id") || return false
    subscribe!(subject, :dashboard, id -> "dashboard:$id") || return false
    !subscribe!(subject, :audit, id -> "duplicate:$id") || return false

    first_delivery = publish(subject, 42)
    first_delivery == Dict(:audit => "audit:42", :dashboard => "dashboard:42") || return false

    unsubscribe!(subject, :dashboard) || return false
    !unsubscribe!(subject, :dashboard) || return false

    publish(subject, 43) == Dict(:audit => "audit:43")
end

end

if abspath(PROGRAM_FILE) == @__FILE__
    ObserverExample.example_passes() || error("Julia Observer verification failed")
    println("Julia Observer: passed")
end
